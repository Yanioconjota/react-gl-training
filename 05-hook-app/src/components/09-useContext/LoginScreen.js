import React, { useContext } from "react";
import { UserContext } from "./UserContext";

export const LoginScreen = () => {
  const { setUser } = useContext(UserContext);

  const loggedUser = {
    id: 666,
    name: 'Homero',
    lastname: 'Thompson',
    email: 'homerothompson@email.com'
  };

  return (
    <>
      <h1>Login screen</h1>
      <hr />
      <button
        className="btn btn-primary"
        onClick={()=> setUser(loggedUser)}>Login</button>
    </>
  );
};
