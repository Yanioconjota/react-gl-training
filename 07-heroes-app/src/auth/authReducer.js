import { types } from "../types/types";

//ejemplo de state
/* const state = {
  name: 'Homero',
  logged: true
}; */

//ejemplo de action
/* const loginAction = {
  type: types.login,
  payload: {
    name: 'Homero',
    email: 'homero@email.com'
  }
}; */

export const authReducer = (state = {}, action) => {
  switch (action.type) {
    case types.login:
      return {
        ...action.payload,
        logged: true
      }
    
    case types.logout:
      return {
        logged: false
      }
  
    default:
      return state;
  }
}