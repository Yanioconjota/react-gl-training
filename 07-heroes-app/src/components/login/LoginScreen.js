import React from "react";
import { useNavigate } from "react-router-dom";

export const LoginScreen = () => {

  const navigate = useNavigate();

  const handleLogin = () => {
    //como 2 argumento recibe opciones, donde replace reemplaza esta ruta para no poder volver con el go back del browser
    navigate('/marvel', {
      replace: true
    });
  };

  return (
    <div className="container-fluid mt-3">
      <div className="row">
        <div className="col-xs-12">
          <h1>Login</h1>
          <hr />
          <button className="btn btn-primary" onClick={handleLogin}>
            Login
          </button>
        </div>
      </div>
    </div>
  );
};
