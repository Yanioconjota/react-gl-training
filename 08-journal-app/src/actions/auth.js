import Swal from "sweetalert2";
import { types } from "../types/types";
import {
  googleAuthProvider,
  getAuth,
  signInWithPopup,
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  signOut,
  updateProfile,
} from "../firebase/firebase-config";
import { finishLoading, startLoading } from "./ui";
import { notesLogout } from "./notes";

//gracias al midleware thunk-redux tenemos acceso al dispatch dentro del return
export const startLoginWithEmailPassword = (email, password) => {
  return (dispatch) => {
    const auth = getAuth();
    dispatch(startLoading());
    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((err) => {
        dispatch(finishLoading());
        //mensaje de error en modal
        Swal.fire('Error', err.message, 'error');
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    const auth = getAuth();
    dispatch(startLoading());
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        //update profile agrega el displayName, ya que este solo se crea cuando nos registramos con google o cualquier red social
        await updateProfile(user, { displayName: name });
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((err) => {
        dispatch(finishLoading());
        //mensaje de error en modal
        Swal.fire("Error", err.message, "error");
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    const auth = getAuth();
    await signOut(auth);
    dispatch(logout());
    dispatch(notesLogout());
  };
};

export const logout = () => ({
  type: types.logout,
});
