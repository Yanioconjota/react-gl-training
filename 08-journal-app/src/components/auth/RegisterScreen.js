import React from 'react';
import { Link } from "react-router-dom";
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

  const [formValues, handleInputChange] = useForm({
    name: 'Homero',
    email: "homerothompson@gmail.com",
    password: "123456",
    password2: "123456"
  });

  const { name, email, password, password2 } = formValues;

  const handleRegister = (e) => {
    e.preventDefault();
    console.log(name, email, password, password2);
  }

  return (
    <>
      <h3 className="auth__title">Register</h3>
      <form onSubmit={ handleRegister }>
        <input type="text"
               placeholder="Name"
               name="name"
               value={ name }
               onChange={ handleInputChange }
               className="auth__input"
               autoComplete="off"/>

        <input type="text"
               placeholder="Email"
               name="email"
               value={ email }
               onChange={ handleInputChange }
               className="auth__input"
               autoComplete="off"/>

        <input type="password" 
               placeholder="Password"
               name="password"
               value={ password }
               onChange={ handleInputChange }
               className="auth__input"/>

        <input type="password" 
               placeholder="Confirm password"
               name="password2"
               value={ password2 }
               onChange={ handleInputChange }
               className="auth__input"/>

        <button type="submit"
                className="btn btn-primary btn-block mb-5">Login
        </button>
        <Link className="link" to="/auth/login">
          Already registered?
        </Link>
      </form>
    </>
  )
}
