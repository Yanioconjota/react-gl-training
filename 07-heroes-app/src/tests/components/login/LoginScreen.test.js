import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { AuthContext } from "../../../auth/authContext";
import { LoginScreen } from "../../../components/login/LoginScreen";
import { types } from "../../../types/types";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas con LoginScreen', () => {

  const contextValue = {
    user: {
      logged: false,
    },
    dispatch: jest.fn(),
  };

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      {/* Permite hacer pruebas en componentes que hacen uso del useNavigate */}
      <MemoryRouter initialEntries={["/login"]}>
        <Routes>
          <Route path="/login" element={<LoginScreen />} />
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );

  test("debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe llamar el login, el navigate y el dispatch con los parÃ¡metros correctos", () => {
    const handleClick = wrapper.find("button").prop("onClick");

    handleClick();

    expect(contextValue.dispatch).toHaveBeenCalledWith({
      payload: { name: 'Homero Thompson' },
      type: types.login,
    });

    expect(mockNavigate).toHaveBeenCalledWith("/marvel", { replace: true });

    localStorage.setItem('lastPath', '/dc');

    handleClick();

    expect(mockNavigate).toHaveBeenCalledWith("/dc", { replace: true });
  });
});