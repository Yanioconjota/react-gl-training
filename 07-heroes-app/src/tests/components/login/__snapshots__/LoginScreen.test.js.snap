// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con LoginScreen debe mostrarse correctamente 1`] = `
<div
  className="container-fluid mt-3"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12"
    >
      <h1>
        Login
      </h1>
      <hr />
      <button
        className="btn btn-primary"
        onClick={[Function]}
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
